### PR TITLE
Add info transport endpoint

### PR DESCRIPTION
**Pull request by lloeki**
**What does this PR do?**

Implement the `/info` endpoint without impacting the current trace endpoint.

**Motivation**

The goal of this commit is twofold:

- prepare for endpoint and configuration negotiation with the agent
- lay ground for reaching to other endpoints than the trace API

To that end, it:

- duplicates and alters a few files, noting the issues encountered towards full reusability of the current transport implementation.
- implements requesting to the `GET /info` endpoint
- does not implement processing the response nor any form of actual negotiation using that response

**Additional Notes**

This may be easier to review by comparing between `lib/ddtrace` and `lib/datadog/core`, e.g:

```
diff --color -w -u lib/{ddtrace,datadog/core}/transport/http/builder.rb
```

The short-term goal is to enable implementing additional endpoints, such as `/v0.7/config` for remote configuration.

Many comments have been dropped in the code itself to act as persistent analysis towards a possible later longer-term refactoring.

**How to test the change?**

The code is isolated and uncalled, so CI should stay green.

Manual testing:

- Step 1: start the agent
```
docker run --rm -it -e DD_APM_ENABLED=true -e DD_ENV=${YOUR_ENV} -e DD_HOSTNAME=${YOUR_HOSTNAME} -e DD_BIND_HOST=0.0.0.0 -e DD_APM_NON_LOCAL_TRAFFIC=true -e DD_API_KEY=${YOUR_API_KEY}-e DD_REMOTE_CONFIGURATION_ENABLED=true -p 8126:8126 datadog/agent:7.42.0
```
- Step 2: hit it with the following code

```
require 'ddtrace'

Datadog.configure do |c|
    c.agent.host = "#{YOUR_AGENT_HOST}"
    # c.diagnostics.debug = true
end

transport_options = {
  agent_settings: Datadog::Core::Configuration::AgentSettingsResolver.call(Datadog.configuration)
}

require 'datadog/core/transport/http'

transport_root = Datadog::Core::Transport::HTTP.root(**transport_options.dup)

puts transport_root.send_info.inspect
```

### CODE REVIEW

**Comment by lloeki** [comment]
Pending:

- specs
- RBS typing

File: lib/datadog/core/transport/http/api.rb [hunk 1]
```diff
@@ -0,0 +1,47 @@
+# frozen_string_literal: true
+
+require_relative '../../encoding'
+
+require_relative '../../../../ddtrace/transport/http/api/map'
+
+# TODO: Decouple standard transport/http/api/instance
+#
+# Separate classes are needed because transport/http/trace includes
```
Comment (GustavoCaso) [inline lib/datadog/core/transport/http/api.rb:9]:
```suggestion
# Separate classes are needed because transport/http/traces includes
```

**Review by GustavoCaso — COMMENTED** [review]
I tested the changes suing Loic script 🎉 

```
bundle exec ruby test.rb
send info
#<Datadog::Core::Transport::HTTP::Negotiation::Response:0x00000001025bb988 @http_response=Datadog::Transport::HTTP::Adapters::Net::Response ok?:true unsupported?:false, not_found?:false, client_error?:false, server_error?:false, internal_error?:, payload:{
	"version": "7.42.0",
	"git_commit": "1b72bcf",
	"endpoints": [
		"/v0.3/traces",
		"/v0.3/services",
		"/v0.4/traces",
		"/v0.4/services",
		"/v0.5/traces",
		"/v0.7/traces",
		"/profiling/v1/input",
		"/telemetry/proxy/",
		"/v0.6/stats",
		"/v0.1/pipeline_stats",
		"/evp_proxy/v1/",
		"/evp_proxy/v2/",
		"/evp_proxy/v3/",
		"/debugger/v1/input",
		"/dogstatsd/v1/proxy",
		"/dogstatsd/v2/proxy",
		"/v0.7/config",
		"/config/set"
	],
	"feature_flags": [
		""
	],
	"client_drop_p0s": true,
	"span_meta_structs": true,
	"long_running_spans": true,
	"config": {
		"default_env": "gustavo.dev",
		"target_tps": 10,
		"max_eps": 200,
		"receiver_port": 8126,
		"receiver_socket": "",
		"connection_limit": 0,
		"receiver_timeout": 0,
		"max_request_bytes": 52428800,
		"statsd_port": 8125,
		"max_memory": 500000000,
		"max_cpu": 0.5,
		"analyzed_spans_by_service": {},
		"obfuscation": {
			"elastic_search": false,
			"mongo": false,
			"sql_exec_plan": false,
			"sql_exec_plan_normalize": false,
			"http": {
				"remove_query_string": false,
				"remove_path_digits": false
			},
			"remove_stack_traces": false,
			"redis": false,
			"memcached": false
		}
	}
}, http_response:#<Net::HTTPOK:0x0000000105c30838>, @version="7.42.0", @endpoints=["/v0.3/traces", "/v0.3/services", "/v0.4/traces", "/v0.4/services", "/v0.5/traces", "/v0.7/traces", "/profiling/v1/input", "/telemetry/proxy/", "/v0.6/stats", "/v0.1/pipeline_stats", "/evp_proxy/v1/", "/evp_proxy/v2/", "/evp_proxy/v3/", "/debugger/v1/input", "/dogstatsd/v1/proxy", "/dogstatsd/v2/proxy", "/v0.7/config", "/config/set"], @config={:default_env=>"gustavo.dev", :target_tps=>10, :max_eps=>200, :receiver_port=>8126, :receiver_socket=>"", :connection_limit=>0, :receiver_timeout=>0, :max_request_bytes=>52428800, :statsd_port=>8125, :max_memory=>500000000, :max_cpu=>0.5, :analyzed_spans_by_service=>{}, :obfuscation=>{:elastic_search=>false, :mongo=>false, :sql_exec_plan=>false, :sql_exec_plan_normalize=>false, :http=>{:remove_query_string=>false, :remove_path_digits=>false}, :remove_stack_traces=>false, :redis=>false, :memcached=>false}}>
```

**Review by GustavoCaso — APPROVED** [review]
LGTM. You can ping me once the tests and RBS files are updated and I would happy give it another pass

**Pull request merged**